# <center>Объявления о продаже автомобилей в Германии<center>

Вам представлен датасет с объявлениями о продаже в Германии на 2022 год. Необходимо осмотреть первично данные и выполнить их предобработку, также сделать соответствующие выводы по полученной информации

### Обзор данных

* `mileage` — Пройденное расстояние на автомобиле
* `make` — Марка автомобиля
* `model` — Модель автомобиля
* `fuel` — Тип топлива
* `gear` — Тип коробки передач
* `offerType` — Тип предложения, то есть «Новые», «Демонстрационные» машины и т.д.
* `price` — Цена продажи автомобиля в объявлении (в долларах)
* `hp` — Количество лошадиных сил
* `year` — Год производства автомобиля

Импортируйте необходимые библиотеки

In [75]:
import pandas as pd
import numpy as np

from pprint import pprint

### Загрузка данных

1. Загрузите необходимый датасет в переменную

In [76]:

cars_df = pd.read_csv('../datasets/germany_cars.csv') # Исходный DataFrame

my_cars_df = cars_df.copy() # DataFrame c которым буду работать

my_cars_df

mileage        make   model               fuel    gear       offerType  \
0       235000         BMW     316             Diesel  Manual            Used   
1        92800  Volkswagen    Golf           Gasoline  Manual            Used   
2       149300        SEAT    Exeo           Gasoline  Manual            Used   
3        96200     Renault  Megane           Gasoline  Manual            Used   
4       156000     Peugeot     308           Gasoline  Manual            Used   
...        ...         ...     ...                ...     ...             ...   
46400       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   
46401       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   
46402       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   
46403       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   
46404       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   

         price     hp  year  
0       6800.0  116.0  2011  
1       6877.0  122.0  2011  
2       6900.0  160.0  2011  
3       6950.0  110.0  2011  
4       6950.0  156.0  2011  
...        ...    ...   ...  
46400  12990.0   71.0  2021  
46401  12990.0   71.0  2021  
46402  12990.0   71.0  2021  
46403  12990.0   71.0  2021  
46404  12990.0   71.0  2021  

[46405 rows x 9 columns]

2. Проведите первичный осмотр данных при помощи известных вам методов(дополнительно можно использовать метод hist() для оценки распределений). Сделайте соответствующие выводы на основе полученной информации

In [77]:
# Поверхностная информация о данном DataFrame

# my_cars_df.info()                                                      # Информация о DataFrame
# hist = correct_cars_df.hist(column=['mileage', 'price', 'hp', 'year']) # В прошлом уроке сдавал

size_colomn = my_cars_df.shape                                          # Размернось DataFrame
count_colmn = size_colomn[1]                                            # Количество колонок
count_row = size_colomn[0]                                              # Количество строк

row_index = my_cars_df.index                                            # Строковые индексы
colomn_index = my_cars_df.columns                                       # Название колонок

colomn_is_nan = my_cars_df.columns[my_cars_df.isnull().any()]           # Колонки в которых есть пропуски

colomn_type = {i: my_cars_df[i].dtype for i in my_cars_df.columns}      # Колонка: тип данных

print(f"Размернось DataFrame: {size_colomn}\n\n"
      f"Количество колонок: {count_colmn}\n\n"
      f"Количество строк: {count_row}\n\n"
      f"Строковые индексы: {row_index}\n\n"
      f"Название колонок: {colomn_index}\n\n"
      f"Колонки в которых есть пропуски: {colomn_is_nan}\n\n")

pprint(colomn_type)

Размернось DataFrame: (46405, 9)

Количество колонок: 9

Количество строк: 46405

Строковые индексы: RangeIndex(start=0, stop=46405, step=1)

Название колонок: Index(['mileage', 'make', 'model', 'fuel', 'gear', 'offerType', 'price', 'hp',
       'year'],
      dtype='object')

Колонки в которых есть пропуски: Index(['model', 'fuel', 'gear', 'price', 'hp'], dtype='object')


{'fuel': dtype('O'),
 'gear': dtype('O'),
 'hp': dtype('float64'),
 'make': dtype('O'),
 'mileage': dtype('int64'),
 'model': dtype('O'),
 'offerType': dtype('O'),
 'price': dtype('float64'),
 'year': dtype('int64')}


3. Переименуйте столбец `offerType` под стандартное название

In [78]:
my_cars_df.rename(columns={'offerType': 'offertype'}, inplace=True)

### Работа с дубликатами данных

4. Необходимо вывести на экран количество дублированных строк в данных, если таковые имеются, а также избавиться от них. Не забудьте сбросить индекс в датафрейме при необходимости.

In [79]:
count_dubl = my_cars_df.duplicated().sum()    # Считаем количество дублирующих строк
my_cars_df.drop_duplicates(inplace=True)      # Удаляем дублирующие строки

my_cars_df.reset_index(inplace=True)          # Сбрасываем индексы до стандартных
del my_cars_df['index']

print(f"Количество дублированных строк: {count_dubl}")  # 2003 , При повторном запуске будет 0. Данные уже перезаписанны

Количество дублированных строк: 2003


### Работа с пропусками

5. Выведите на экран количество пропусков в каждом столбце, опишите их, почему они могут возникнуть и как их возможно исправить

In [80]:
check_colomn_is_nan = my_cars_df.isna().sum()
only_colomn_is_nan = check_colomn_is_nan[check_colomn_is_nan > 0]

print(only_colomn_is_nan)

model     137
fuel      247
gear      175
price    1520
hp         24
dtype: int64


6. Заполните пропуски в `model`,`fuel`,`gear`

In [81]:
my_cars_df['model'] = my_cars_df['model'].fillna('Модель не указана')
my_cars_df['fuel'] = my_cars_df['fuel'].fillna('92/95/100')
my_cars_df['gear'] = my_cars_df['gear'].fillna('AT/MT')


## Работа с пропусками данных ПРО (желательно выполнить, но необязательно)

7. Необходимо заполнить пропуски в столбце `price` с помощью медианного значения по моделям автомобиля.

In [82]:
median_model = my_cars_df.groupby('model')['price'].median()  # Находим медиану по каждой модели

my_cars_df['price'] = my_cars_df.apply(lambda row: median_model[row['model']] if pd.isna(row['price']) else row['price'], axis=1) # Заполняем

my_cars_df.loc[7] # Проверка на одной конкретной строчки

my_cars_df[my_cars_df['price'].isna()]  # Не могу понять почему ути строки не заполнились (

my_cars_df['price'] = my_cars_df['price'].fillna(10000) # Заполнил две строчки в ручную

## Продолжение базовых заданий

7. Если предыдущее задание показалось Вам сложным, то удалите оставшиеся пропуски.

In [83]:
print(f"Оствшееся пропуски:\n\n{my_cars_df.isna().sum()}")

my_cars_df['hp'] = my_cars_df['hp'].fillna(0) # Заполнил нулями


Оствшееся пропуски:

mileage       0
make          0
model         0
fuel          0
gear          0
offertype     0
price         0
hp           24
year          0
dtype: int64


Проверьте, все ли пропуски точно удалились из датафрейма

In [84]:
print(f"Оствшееся пропуски:\n\n{my_cars_df.isna().sum()}")

Оствшееся пропуски:

mileage      0
make         0
model        0
fuel         0
gear         0
offertype    0
price        0
hp           0
year         0
dtype: int64


8. Оцените количество уникальных марок автомобилей (производителей) с помощью метода unique() и выведите их количество числом (пожалуйста, не считайте вручную)

In [85]:
count_make = my_cars_df['make'].unique().size

print(f'Кол-во уникальных марок: {count_make}')

Кол-во уникальных марок: 77


9. Выведите топ 20 марок автомобилей (производителей) по количеству объявлений с помощью метода value_counts()

In [138]:
top_make = my_cars_df['make'].value_counts().head(20)
top_make

make
Volkswagen       6725
Opel             4616
Ford             4290
Skoda            2755
Renault          2704
Audi             2559
BMW              2373
Mercedes-Benz    2281
SEAT             1834
Hyundai          1758
Fiat             1504
Peugeot          1182
Toyota           1149
Kia               968
smart             937
Citroen           924
Volvo             778
Nissan            730
Mazda             699
Dacia             666
Name: count, dtype: int64

### Группировка данных

10. Сгруппируйте данные по столбцу `make`, выведите для каждого производителя среднюю стоимость их автомобиля, а также среднее число лошадиных сил и отсортируйте по убыванию столбца `price`. Отобразите только топ 20 самых дорогих производителей. Напишите вывод, всё ли в порядке с полученными данными или Вас что-то смущает

In [87]:
group_make = my_cars_df.groupby(by='make').agg(price_mean=('price','mean'), hp_mean=('hp', 'mean'))
group_make = group_make.nlargest(20, 'price_mean')
group_make

price_mean     hp_mean
make                                  
Maybach      450479.333333  543.000000
Ferrari      324028.181818  693.909091
Lamborghini  305698.777778  690.111111
McLaren      207538.181818  646.636364
Bentley      182383.966667  554.866667
Aston        163442.960000  536.200000
Rolls-Royce  133266.666667  605.000000
Corvette     110160.000000  687.000000
Porsche       90140.474138  412.383621
FISKER        69900.000000  408.000000
Maserati      69118.818182  376.363636
Alpina        66046.300000  443.000000
Polestar      64150.000000  408.000000
Tesla         62392.913043  491.478261
Morgan        61950.000000  177.000000
Alpine        53783.000000  244.800000
Land          51825.472393  250.288344
RAM           47945.000000  401.000000
Cadillac      45032.142857  348.285714
Dodge         44661.043478  395.869565

11. Создайте новый столбец из `make` и `model` и выведите топ 10 самый дорогих и дешевых автомобилей с помощью средней стоимости

In [88]:
my_cars_df['make_model'] = my_cars_df['make'].str.cat(my_cars_df['model'], sep='-')
price_df = my_cars_df.groupby(by='make_model').agg(price_mean=('price', 'mean')).sort_values('price_mean', ascending=False)
price_df['price_mean'] = round(price_df['price_mean'])


display(price_df.head(10))
display(price_df.tail(10))

price_mean
make_model                       
Ferrari-F12             1199900.0
Mercedes-Benz-S 650      717078.0
Maybach-Pullman          608439.0
Ferrari-812              439900.0
Aston-Martin             398000.0
Lamborghini-Aventador    371423.0
McLaren-675LT            297900.0
Ferrari-F8 Tributo       294417.0
Bentley-Flying Spur      251560.0
McLaren-720S             251145.0

price_mean
make_model                                    
Daihatsu-Cuore                          3495.0
Fiat-Linea                              3400.0
Peugeot-206                             3366.0
Volkswagen-Fox                          3011.0
Hyundai-Getz                            2800.0
Trailer-Anhänger-Модель не указана      2750.0
Volvo-Модель не указана                 2499.0
Others-Landwind                         1850.0
Brilliance-BS4                          1500.0
Lada-Priora                             1190.0

## Задания ПРО (желательно выполнить, но необязательно)

12. Выведите информацию о DataFrame с помощью метода describe() для категориальных данных , кратко опишите выведенную информацию

In [89]:
object_category = my_cars_df.describe(include='object')

int_category = my_cars_df.describe(include='int64')

float_category = my_cars_df.describe(include='float64')

print(object_category, int_category, float_category, sep='\n\n')

              make  model      fuel    gear offertype       make_model
count        44402  44402     44402   44402     44402            44402
unique          77    842        12       4         5              884
top     Volkswagen   Golf  Gasoline  Manual      Used  Volkswagen-Golf
freq          6725   1458     27311   28920     38935             1458

            mileage          year
count  4.440200e+04  44402.000000
mean   7.221993e+04   2015.936061
std    6.237677e+04      3.121481
min    0.000000e+00   2011.000000
25%    2.131350e+04   2013.000000
50%    6.105050e+04   2016.000000
75%    1.060000e+05   2019.000000
max    1.111111e+06   2021.000000

              price            hp
count  4.440200e+04  44402.000000
mean   1.649862e+04    133.493379
std    1.937145e+04     75.396138
min    1.100000e+03      0.000000
25%    7.499000e+03     86.000000
50%    1.099000e+04    116.000000
75%    1.946675e+04    150.000000
max    1.199900e+06    850.000000


13. Выведите на экран все BMW с пробегом меньше 1000 миль и годом производства после 2020 включительно, то есть, новые автомобили. Какие выводы можно сделать?

In [90]:
my_cars_df[(my_cars_df['make'] == 'BMW') & (my_cars_df['mileage'] < 1000) & (my_cars_df['year'] >= 2020)]

mileage make model               fuel       gear       offertype  \
1587        50  BMW   320           Gasoline  Automatic            Used   
5665        25  BMW   330  Electric/Gasoline  Automatic   Demonstration   
5670        25  BMW    X5  Electric/Gasoline  Automatic   Demonstration   
10239       41  BMW   118           Gasoline     Manual  Employee's car   
15032       10  BMW    X2           Gasoline  Automatic  Pre-registered   
18837      789  BMW   530           Gasoline  Automatic  Employee's car   
27614      500  BMW    X5             Diesel  Automatic   Demonstration   
29229        5  BMW   730             Diesel  Automatic   Demonstration   
29235      600  BMW  X5 M           Gasoline  Automatic   Demonstration   
31782      250  BMW    X1           Gasoline     Manual   Demonstration   
42964      101  BMW    M4           Gasoline  Automatic   Demonstration   
44339       45  BMW   iX3           Electric      AT/MT            Used   
44356      100  BMW   340           Gasoline  Automatic   Demonstration   

         price     hp  year make_model  
1587   33950.0  184.0  2020    BMW-320  
5665   54680.0  292.0  2021    BMW-330  
5670   89980.0  394.0  2021     BMW-X5  
10239  24960.0  136.0  2021    BMW-118  
15032  41450.0  178.0  2021     BMW-X2  
18837  51990.0  252.0  2020    BMW-530  
27614  81995.0  286.0  2021     BMW-X5  
29229  89995.0  286.0  2021    BMW-730  
29235  71800.0  625.0  2021   BMW-X5 M  
31782  29260.0  136.0  2021     BMW-X1  
42964  64450.0  450.0  2020     BMW-M4  
44339  58000.0  286.0  2021    BMW-iX3  
44356  82648.0  374.0  2021    BMW-340

14. Напишите функцию, которая будет применена к столбцу `price` с помощью apply(). Она должна категоризировать машины по данному признаку на дорогие, средней ценовой категории и дешевые. (границы категорий подберите самостоятельно). Дополнительно выведите количество автомобилей в каждой новой категории.

In [91]:
def category_price(row):
    '''Дешевая цена/Средняя цена/Дорогая цена.'''

    if row['price'] < 10000:
        return 'Дешевая цена'
    if 10000 < row['price'] < 30000:
        return 'Средняя цена'
    if row['price'] > 30000:
        return 'Дорогая цена'


my_cars_df['category_price'] = my_cars_df.apply(category_price, axis=1)

count_cheap = my_cars_df[my_cars_df['category_price'] == 'Дешевая цена'].index.size
count_medium = my_cars_df[my_cars_df['category_price'] == 'Средняя цена'].index.size
count_expensive = my_cars_df[my_cars_df['category_price'] == 'Дорогая цена'].index.size

print(f"Дешевых на рынке: {count_cheap}\n"
      f"Средней стоимости на рынке: {count_medium}\n"
      f"Дорогих на рынке: {count_expensive}")


Дешевых на рынке: 20096
Средней стоимости на рынке: 19209
Дорогих на рынке: 5022


15. Сделайте общий вывод о проделанной работе

Мною был проанализированы данные о продажах автомобилей в Германии с помощью библиотеки **pandas**

- Количество объявлений о продажах автомобилей c 2011 по 2022 год **46405** из них актуальных только **44401**
- Разнообразие рынка **77** разных марок автомобилей
- Самая популярная марка автомобиля на продаже: **Volkswagen**
- Самые не леквидные:  **Trucks-Lkw**, **DAF**, **Zhidou**, **9ff**, **Brilliance**, **FISKER**, **Tazzari**, **Isuzu**
- Средний пробег на продаваемые автомобили около **72219**
- Самая дешевая машина продавалась в **2011** году , это был **Citroen С1** с пробегом **204000** , стоимостью **1100.0** 
- Самая дорогая машина продавалась в **2017** году , это был **Ferrari F12** с пробегом **431** , стоимостью **1199900.0** 
- Средняя стоимость автомобиля в эти года составляла **16498**
- В среднем цена автомобиля росла с динамикой **+2150** , орентируясь на среднюю цену автомобиля
- Лидеры по количеству объявлений по годам: 
    - С 2011 по 2020 года беспорным лидером по продажам автомобилей был **Volkswagen**
    - На 2021 лидер **Skoda**
- Состояния автомобилей на рынке:
    - Used (Б/у): 38935
    - Demonstration (Демонстрационное): 2256
    - Pre-registered (Предрегистрационное): 2114
    - Employee's car (Служебный автомобиль): 1086
    - New (Новый): 11

